src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs

```diff
@@ -69,25 +72,73 @@ internal partial class PackageInstallerService : AbstractDelayStartedService, IP
             IThreadingContext threadingContext,
             VisualStudioWorkspaceImpl workspace,
             SVsServiceProvider serviceProvider,
-            IVsEditorAdaptersFactoryService editorAdaptersFactoryService)
+            IVsEditorAdaptersFactoryService editorAdaptersFactoryService,
+            [Import(AllowDefault = true)] Lazy<IVsPackageInstallerServices> packageInstallerServices,
+            [Import(AllowDefault = true)] Lazy<IVsPackageInstaller2> packageInstaller,
+            [Import(AllowDefault = true)] Lazy<IVsPackageUninstaller> packageUninstaller,
+            [Import(AllowDefault = true)] Lazy<IVsPackageSourceProvider> packageSourceProvider)
             : base(threadingContext, workspace, SymbolSearchOptions.Enabled,
                               SymbolSearchOptions.SuggestForTypesInReferenceAssemblies,
                               SymbolSearchOptions.SuggestForTypesInNuGetPackages)
         {
             _workspace = workspace;
             _serviceProvider = serviceProvider;
             _editorAdaptersFactoryService = editorAdaptersFactoryService;
+            _packageInstallerServices = packageInstallerServices;
+            _packageInstaller = packageInstaller;
+            _packageUninstaller = packageUninstaller;
+            _packageSourceProvider = packageSourceProvider;
         }
 
-        public ImmutableArray<PackageSource> PackageSources { get; private set; } = ImmutableArray<PackageSource>.Empty;
+        public async ValueTask<ImmutableArray<PackageSource>> GetPackageSourcesAsync(CancellationToken cancellationToken)
+        {
+            var packageSources = _packageSources;
+            if (packageSources != null)
+            {
+                return packageSources;
+            }
+
+            try
+            {
+                await ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+                cancellationToken.ThrowIfCancellationRequested();
+
+                packageSources = _packageSourceProvider.Value.GetSources(includeUnOfficial: true, includeDisabled: false)
```

Comment:
@rrelyea Do the NuGet APIs provide both of the following?

1. `IVsPackageSourceProvider.GetSources` can be called from a background thread
2. `IVsPackageSourceProvider.GetSources` is guaranteed to complete even if the UI thread is *blocked* (i.e. no main thread dependency)

If so, I can remove the use of `SwitchToMainThreadAsync` and unblock this pull request.

Reply:
so point of this whole service is so that we can call this in code fix which Wait() on UI thread, wont UI thread switch cause dead lock here?

Reply:
It's safe to call from a BG thread.

Reply:
@nkolev92 I need an answer to the second question as well.

Reply:
The method itself doesn't need the UI thread.
It would've probably deadlocked already if it was using it.

Reply:
I'll update this pull request accordingly.

Reply:
@sharwell @nkolev92 It seems the change here indeed caused a deadlock: https://github.com/dotnet/roslyn/issues/39889. It also seems very likely to incur UI delays, invoking an external service from a code fix that blocks the user is not ideal.

Reply:
I'd create an issue on NuGet/Home for this. 
NuGet has seen many delays with our synchronous services (and obviously deadlocks like now), so we are pushing for an asynchronous rewrite.

Reply:
It'd be great if you can provide repro steps for the hang. 
That way we can validate any fixes. We should also look if there's any tel tracking these hangs.../cc @zivkan

Reply:
I have updated the description in https://github.com/dotnet/roslyn/issues/39889#issue-525249832 with the repro steps, which seems to give me pretty consistent deadlock. I can file a separate issue on NuGet/Home for the same, as we probably want to keep https://github.com/dotnet/roslyn/issues/39889 to change Roslyn code to not invoke NuGet APIs from inside a code fixer to avoid potential UI delays.